### PR TITLE
sql: implement idle_in_session_timeout setting to cancel sessions

### PR DESCRIPTION
**Pull request by RichardJCai**
sql: implement idle_in_session_timeout setting to cancel sessions 
that exceed a given idle threshold.

Release note (sql change): Add support for idle_in_session_timeout
variable to allow automatically terminating sessions that idle past
a certain threshold.

Release note (sql change): Added cluster setting sql.defaults.idle_in_session_timeout
for setting a default idle_in_session_timeout value for new sessions.
This gives users the ability to set idle_in_session_timeout values for all
newly created sessions.

We've decided not to let being able to send an notice message block us for this, created separate issue about being able to send a notice when cancelling a session.
https://github.com/cockroachdb/cockroach/issues/51237

Fixes #https://github.com/cockroachdb/cockroach/issues/50706, #36054

### CODE REVIEW

**Comment by RichardJCai** [comment]
Looking for some suggestions on how to send the client a message before cancelling the session. Currently it's unclear why the session dies.
(Addressing the TODO from line 2247 of conn_executor.go)

**Comment by rohany** [comment]
> Looking for some suggestions on how to send the client a message before cancelling the session. Currently it's unclear why the session dies.

You could use the noticeSender to send a notice

**Comment by RichardJCai** [comment]
> > Looking for some suggestions on how to send the client a message before cancelling the session. Currently it's unclear why the session dies.
> 
> You could use the noticeSender to send a notice

Currently noticeSender only sends a message when the command finishes, in this case the session gets closed before the command "finishes" maybe we can defer the actually killing of the session until the command finishes?

**Comment by otan** [comment]
> currently noticeSender only sends a message when the command finishes

https://github.com/cockroachdb/cockroach/pull/49190

File: pkg/sql/conn_executor.go [hunk 1]
```diff
@@ -1047,6 +1048,10 @@ type connExecutor struct {
 		// LastActiveQuery contains a reference to the AST of the last
 		// query that ran on this session.
 		LastActiveQuery tree.Statement
+
+		// IdleInSessionTimeout is returned by the AfterFunc call that cancels the
+		// session if the idle time exceeds the idle_in_session_timeout.
+		IdleInSessionTimeout *time.Timer
```
Comment (knz) [inline pkg/sql/conn_executor.go:1054]: nit: if you create an interface (your own struct) here, you can make it have a `Stop()` method that's always available, this way you can avoid `if xxx != nil` all over the place.

**Comment by awoods187** [comment]
Can you tie this PR to the two open issues to track progress?

**Review by solongordon — COMMENTED** [review]
Nice. Could you also add `sql.defaults.idle_in_session_timeout`? This is not so useful if it has to be set on every session.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51223#-:-MC7nwZt8keuOGn4g822:b7r6n75)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RichardJCai, and @solongordon)

**Review by solongordon — COMMENTED** [review]
Actually I wonder if this should be a cluster setting only. Not sure if it makes sense to let sessions override it. I'll inquire on the ticket.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51223#-:-MC7p9T8ARsby4G3N_si:b-xugxhm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @RichardJCai, and @solongordon)

**Comment by knz** [comment]
> Actually I wonder if this should be a cluster setting only. 

FWIW:

- making it consistent with other vars (which already have both) might trump the specific user demand. (Consistent means also easier to explain/understand)

- For testing purposes I believe it's going to be necessary to have the manual SET otherwise the test is harder to write.

**Comment by RichardJCai** [comment]
> Nice. Could you also add `sql.defaults.idle_in_session_timeout`? This is not so useful if it has to be set on every session.
> 
> >

I actually have the implementation done for this as well, I wasn't sure whether to include this in the PR or not. I'll go ahead and add it in here.

**Review by solongordon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51223)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @RichardJCai)

___
*[pkg/sql/run_control_test.go, line 384 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51223#-MC7xbzdEsqSSr-nDLMr:-MC7xbze1JZSX3S_5rL4:bqoep4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f1f29740e4c668787e208c9e90238ee2bba38a/pkg/sql/run_control_test.go#L384)):*
> ```Go
> 	}
> 
> 	time.Sleep(2 * time.Second)
> ```

Don't love seeing all these sleeps in a test. I wonder if there's a reasonable way to mock out the clock here. Not obvious to me the best way to go about that.

**Review by solongordon — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51223)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @RichardJCai)

___
*[pkg/sql/conn_executor_exec.go, line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51223#-MC82lyL-hMyJCq1jbzI:-MC82lyM7N96LmUGhpGF:b6zckwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f1f29740e4c668787e208c9e90238ee2bba38a/pkg/sql/conn_executor_exec.go#L123)):*
> ```Go
> 		// Cancel the session if the idle time exceeds the idle in session timeout.
> 		ex.mu.IdleInSessionTimeout = time.AfterFunc(
> 			ex.sessionData.IdleInSessionTimeout-timeutil.Since(timeutil.Now()),
> ```

Wait, what is the point of `-timeutil.Since(timeutil.Now())`?

**Review by solongordon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51223)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @RichardJCai)

___
*[pkg/sql/conn_executor_exec.go, line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51223#-MC836WF3wNcuPBOLfAN:-MC836WF3wNcuPBOLfAO:b-ra3vjd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f1f29740e4c668787e208c9e90238ee2bba38a/pkg/sql/conn_executor_exec.go#L126)):*
> ```Go
> 			func() {
> 				ex.cancelSession()
> 			})
> ```

Nit: You can just write `ex.cancelSession`. The wrapper is unnecessary.

**Review by solongordon — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51223)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @RichardJCai)

___
*[pkg/sql/conn_executor_exec.go, line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51223#-MC84Y2hE0uZB894VRUN:-MC84Y2i8R2PK7Mk-TY9:bq4k7cz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f1f29740e4c668787e208c9e90238ee2bba38a/pkg/sql/conn_executor_exec.go#L122)):*
> ```Go
> 	if ex.sessionData.IdleInSessionTimeout > 0 {
> 		// Cancel the session if the idle time exceeds the idle in session timeout.
> 		ex.mu.IdleInSessionTimeout = time.AfterFunc(
> ```

I wonder if anything funky can happen if this runs after the session has already closed. Do we want some cleanup to happen if the conn executor closes before this runs?

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51223)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @solongordon)

___
*[pkg/sql/conn_executor_exec.go, line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51223#-MC84Y2hE0uZB894VRUN:-MC8hedq8AWuZOeLkTpr:b-q0jloy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f1f29740e4c668787e208c9e90238ee2bba38a/pkg/sql/conn_executor_exec.go#L122)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

I wonder if anything funky can happen if this runs after the session has already closed. Do we want some cleanup to happen if the conn executor closes before this runs?
</blockquote></details>

Added a call to Stop when connExecutor is closed to kill the goroutine.
___
*[pkg/sql/conn_executor_exec.go, line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51223#-MC82lyL-hMyJCq1jbzI:-MC8U8wx3D9mvT7jHi9u:b-xbljy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f1f29740e4c668787e208c9e90238ee2bba38a/pkg/sql/conn_executor_exec.go#L123)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Wait, what is the point of `-timeutil.Since(timeutil.Now())`?
</blockquote></details>

Oops I forgot to update this, yeah this should be removed.
___
*[pkg/sql/run_control_test.go, line 384 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51223#-MC7xbzdEsqSSr-nDLMr:-MC8hPML6gnuVlNcf0bQ:b-lxo68k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f1f29740e4c668787e208c9e90238ee2bba38a/pkg/sql/run_control_test.go#L384)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Don't love seeing all these sleeps in a test. I wonder if there's a reasonable way to mock out the clock here. Not obvious to me the best way to go about that.
</blockquote></details>

Yeah I can't think of a way to mock out the clock while still testing the functionality,

File: pkg/sql/conn_executor.go (hunk 1 above)
Comment (RichardJCai) [reply pkg/sql/conn_executor.go]: Done

**Comment by RichardJCai** [comment]
@awoods187 In this implementation, there's a cluster setting but also session setting for idle_in_session_timeout, the cluster setting determines the idle_in_session_timeout value for new sessions but the user can always override it. This is consistent with the behaviour of other variables that can be set at the cluster and session level but I just want to double check that this is okay from a security standpoint for the customer.

**Comment by awoods187** [comment]
I'm checking with the customer but i think we should keep it consistent for now

**Comment by jordanlewis** [comment]
Just to weigh in on the cluster / session setting pairs, I think it's a good idea to keep this paradigm for now, but we should devote serious thought soon to how to provide more control for customers. A cluster setting is extremely broad and affects every session. A session setting requires manual setting on each connection.

Other options we should explore:

1. Database defaults, which would permit creating a default setting for connections to a particular database.
1. Role-based defaults, which would permit a setting to have a default for a particular role (still overridable via session).

We may wish to explore a few options here and create a unified hierarchy of setting defaults.

**Comment by awoods187** [comment]
@jordanlewis its clear that certain customers want role-based views long-term. Consider 
 Oracle. https://docs.oracle.com/cd/B19306_01/network.102/b14266/admusers.htm#i1012785

I think its fine for that to come in a subsequent release though.

**Review by solongordon — APPROVED** [review]
:lgtm: My only complaint is that 11 seconds of sleeps in the test seems like a lot... I wonder if you could cut down on that without risking flakiness.

> Reviewed 2 of 2 files at r2, 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51223#-:-MChhXwjD_Ne27ct_UFz:b-8j5i5x)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)

**Comment by RichardJCai** [comment]
> :lgtm: My only complaint is that 11 seconds of sleeps in the test seems like a lot... I wonder if you could cut down on that without risking flakiness.
> 
> > Reviewed 2 of 2 files at r2, 4 of 4 files at r3.
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51223#-:-MChhXwjD_Ne27ct_UFz:b-8j5i5x)_ status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)

I'll try halving the time and seeing if that introduces any flakiness.

**Comment by RichardJCai** [comment]
Thanks for the reviews!

**Comment by RichardJCai** [comment]
bors r=solongordon

**Pull request merged**